orttraining/orttraining/python/training/ortmodule/_execution_agent.py

```diff
@@ -2,134 +2,327 @@
 
         return IOBinding(self._inference_session)
 
-    def run_forward(self, iobinding, run_options):
-        """
-         Compute the forward graph.
-         :param iobinding: the iobinding object that has graph inputs/outputs bind.
-         :param run_options: See :class:`onnxruntime.RunOptions`.
+    def _process_inputs(self, inputs: Sequence[torch.Tensor]) -> IOBinding:
+        """Runs the forward graph on execution_session with given model inputs and device"""
+
+        # Assert that the input and model device match
+        _utils._check_same_device(self._device, "Input argument to forward", *inputs)
+
+        # TODO: Try to reuse the output buffers as some of the output tensors are same sizes,
+        #   especially the backward graph outputs.
+        # REVIEW(codemzs): Consolidate Training Agent with InferenceAgent on C++ side to not
+        # have the need for passing IOBinding.
+        io_binding = IOBinding(self._inference_session)
+
+
+        # Use IO binding
+        _utils._create_iobinding(io_binding, inputs, self._onnx_model, self._device)
+
+        return io_binding
+
+    def _process_outputs(self, forward_outputs: Sequence[OrtValue]) -> Tuple[Sequence[torch.Tensor], RunStateInfo]:
+        user_outputs = tuple(_utils._ortvalue_to_torch_tensor(forward_output._ortvalue) for forward_output in forward_outputs)
+        state = None
+
+        # Assert that the outputs and model device match
+        _utils._check_same_device(self._device, "Output argument from forward", *user_outputs)
+
+        output_info = [(output.shape, output.device, output.dtype) for output in user_outputs]
+        run_info = RunStateInfo(state, output_info)
+        # Return user outputs and forward run information
+        return user_outputs, run_info
+
+    def _run_forward(self, iobinding: IOBinding, run_options: RunOptions) -> ExecutionAgentOutput:
+        """Computes the forward graph using IOBinding.
+
+        Args:
+            iobinding: the iobinding object that has graph inputs/outputs bind.
+            run_options: See :class:`onnxruntime.RunOptions`.
         """
 
         self._inference_session.run_with_iobinding(iobinding, run_options)
         ortvalues = iobinding.get_outputs()
         return ExecutionAgentOutput(ortvalues)
 
+    def forward(self, *inputs: torch.Tensor) -> Tuple[Sequence[torch.Tensor], RunStateInfo]:
+        """Performs forward computation
+
+        Args:
+            inputs: torch input tensors
+
+        Returns:
+            a tuple of torch output tensors and RunStateInfo object containing forward run information
+        """
+        run_options = RunOptions()
+        io_binding = self._process_inputs(inputs)
+
+        # Run and return module outputs.
+        ort_output = self._run_forward(io_binding, run_options)
+
+        return self._process_outputs(ort_output.ortvalues)
+
+class YieldOpNotFound(Exception):
+    pass
+
+@dataclass(frozen=True)
+class YieldOpInfo:
```

Comment:
My understanding is that this is a refactoring that gathered attributes from the onnx graph related to the YieldOp op. The original code was at `_ORTModuleFunction.backward`. Is that correct?

Reply:
Yes